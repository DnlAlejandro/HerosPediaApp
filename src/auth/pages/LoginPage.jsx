import { useContext } from "react"
import { useNavigate } from "react-router-dom"
import { AuthContext } from "../context/AuthContext"

export const LoginPage = () => {

  const { login } = useContext( AuthContext )
  const navigate = useNavigate()

  const handleLogin = () => {

    const lastpath = localStorage.getItem('lastpath') || '/heroes-app/';

    login('DnlAlejandro');

    navigate(lastpath, {
      replace: true
  })
  }

  return (
    <div className="flex flex-col items-center justify-center h-screen">
      <h1 className="text-3xl font-bold text-gray-900">Login</h1>
      <button
        className="w-full max-w-sm mt-8 bg-indigo-500 hover:bg-indigo-700 rounded-lg shadow-xl font-medium text-white px-4 py-2"
        onClick={handleLogin}
      >
        Join
      </button>
    </div>
  )
}
