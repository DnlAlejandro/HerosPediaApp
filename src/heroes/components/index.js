export { HeroList } from './HeroList';
export { HeroCard } from './HeroCard';