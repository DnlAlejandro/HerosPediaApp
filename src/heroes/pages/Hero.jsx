import { useParams } from "react-router-dom";
import { getHeroeById } from "../helpers";
import { Navigate } from "react-router-dom";
import { useNavigate } from "react-router-dom";
import { useMemo } from "react";

export const Hero = () => {
    const { id } = useParams();

    const { superhero, publisher, alter_ego, first_appearance, characters } =
        useMemo(() => getHeroeById(id), [id]);

    const heroImageURL = `/heroes/${id}.jpg`;

    if (!superhero) {
        return <Navigate to="/marvel" />;
    }

    const navigate = useNavigate();

    const handleNavigateBar = () => {
        navigate(-1, {
            replace: true,
        });
    };

    return (
        <>
            <div className="flex gap-5">
                <div>
                    <img
                        src={heroImageURL}
                        className="rounded-lg h-full max-h-[70vh] min-w-72 hover:scale-105"
                        alt={superhero}
                    />
                    <h1 className="mt-4 font-bold text-3xl text-slate-600">
                        {superhero}
                    </h1>
                    <span
                        className={
                            publisher === "Marvel Comics"
                                ? "text-red-700"
                                : "text-blue-600"
                        }
                    >
                        {publisher}
                    </span>
                </div>
                <div className="w-full">
                    {alter_ego === characters ? (
                        <div className="my-5 shadow-md p-5 rounded-md">
                            <h4 className="font-semibold text-2xl text-indigo-600">
                                Alter Ego & Character
                            </h4>
                            <span className="text-sm text-slate-500">
                                {alter_ego}
                            </span>
                        </div>
                    ) : (
                        <div className="*:my-5 *:shadow-md *:p-5 *:rounded-md">
                            <div>
                                <h4 className="font-semibold text-2xl text-indigo-600">
                                    Alter Ego
                                </h4>
                                <span className="text-sm text-slate-500">
                                    {alter_ego}
                                </span>
                            </div>
                            <div>
                                <h4 className="font-semibold text-2xl text-indigo-600">
                                    Characters
                                </h4>
                                <span className="text-sm text-slate-500">
                                    {characters}
                                </span>
                            </div>
                        </div>
                    )}
                    <div className="shadow-md p-5 rounded-md">
                        <h4
                            className={`font-semibold text-2xl ${
                                publisher === "Marvel Comics"
                                    ? "text-red-700"
                                    : "text-blue-600"
                            }`}
                        >
                            First Appearance
                        </h4>
                        <span className="text-sm text-slate-500">
                            {first_appearance}
                        </span>
                    </div>
                    <button
                        className="bg-indigo-600 mt-4 text-white text-sm w-full h-12 rounded-md hover:bg-indigo-700"
                        onClick={handleNavigateBar}
                        data-testid="goback-btn"
                    >
                        Go back
                    </button>
                </div>
            </div>
        </>
    );
};
