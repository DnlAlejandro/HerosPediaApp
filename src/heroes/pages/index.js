export { DcPage } from './DcPage';
export { MarvelPage } from './MarvelPage';
export { Hero } from './Hero';
export { SearchPage } from './SearchPage';