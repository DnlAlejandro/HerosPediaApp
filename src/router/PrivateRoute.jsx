import { useContext } from "react";
import { AuthContext } from "../auth";
import { Navigate } from "react-router-dom";
import { useLocation } from "react-router-dom";

export const PrivateRoute = ({ children }) => {
    const { logged } = useContext(AuthContext);
    const { pathname, search } = useLocation();

    const lastpath = `${pathname}${search}`;

    localStorage.setItem("lastpath", lastpath);

    return logged ? children : <Navigate to="/login" />;
};
