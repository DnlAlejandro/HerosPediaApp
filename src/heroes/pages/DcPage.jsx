import { HeroList } from "../components"

export const DcPage = () => {
  return (
    <div>
      <span className="text-blue-500">Characters</span>
      <h1 className="text-4xl font-bold mb-3 text-blue-700">DC Comics</h1>

      <HeroList
        publisher={'DC Comics'}
      />

    </div>
  )
}
