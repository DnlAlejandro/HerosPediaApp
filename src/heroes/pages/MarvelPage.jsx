import { HeroList } from "../components";

export const MarvelPage = () => {
    return (
        <div>
            <span className="text-red-500">Characters</span>
            <h1 className="text-4xl font-bold mb-3 text-red-700">
                Marvel Comics
            </h1>

            <HeroList publisher={"Marvel Comics"} />
        </div>
    );
};
