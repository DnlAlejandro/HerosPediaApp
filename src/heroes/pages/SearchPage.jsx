import { useLocation } from "react-router-dom";
import { HeroCard } from "../components";
import queryString from "query-string";
import { getHeroeByName } from "../helpers";

export const SearchPage = () => {
    const location = useLocation();

    const { q = "" } = queryString.parse(location.search);

    const heroes = getHeroeByName(q);

    const searchAlert = (
        <div
            data-testid="search-info"
            className="mt-5 bg-blue-100 p-5 rounded-lg text-blue-600 border border-blue-300/40 flex items-center gap-3 "
        >
            <svg
                xmlns="http://www.w3.org/2000/svg"
                fill="none"
                viewBox="0 0 24 24"
                strokeWidth={1.5}
                stroke="currentColor"
                className="w-6 h-6"
            >
                <path
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    d="m21 21-5.197-5.197m0 0A7.5 7.5 0 1 0 5.196 5.196a7.5 7.5 0 0 0 10.607 10.607Z"
                />
            </svg>
            <p className="font-light">Search a hero</p>
        </div>
    );

    const errorAlert = (
        <div
            data-testid="notfound-info"
            className="mt-5 bg-yellow-100 p-5 rounded-lg text-yellow-600 border border-yellow-300/40 flex items-center gap-3"
        >
            <svg
                xmlns="http://www.w3.org/2000/svg"
                fill="none"
                viewBox="0 0 24 24"
                strokeWidth={1.5}
                stroke="currentColor"
                className="w-6 h-6"
            >
                <path
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    d="M12 9v3.75m9-.75a9 9 0 1 1-18 0 9 9 0 0 1 18 0Zm-9 3.75h.008v.008H12v-.008Z"
                />
            </svg>
            <p className="font-light">
                There are not results with{" "}
                <span className="font-medium">{q}</span>
            </p>
        </div>
    );

    return (
        <>
            <h1 className="text-3xl font-bold text-indigo-600">Results</h1>
            {heroes.length !== 0 && (
                <span className="text-md text-slate-600">{q}</span>
            )}

            {q === "" ? searchAlert : heroes.length === 0 && errorAlert}

            <div className="flex gap-4 flex-wrap">
                {heroes.map((hero) => (
                    <HeroCard
                        key={hero.id}
                        id={hero.id}
                        superhero={hero.superhero}
                    />
                ))}
            </div>
        </>
    );
};
