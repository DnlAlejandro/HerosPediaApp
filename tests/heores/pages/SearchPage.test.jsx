import { render, screen, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";
import { AuthContext } from "../../../src/auth";
import { PrivateRoute } from "../../../src/router/PrivateRoute";
import { HeroesRoutes } from "../../../src/heroes";
import { act } from "react-dom/test-utils";

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate,
    
}))

describe('Prueba en <SearchPage/>', () => {

    
    const contextValue = {
        logged: true,
        name: 'coofcoding'
    }

    beforeEach(() => jest.clearAllMocks());

    test('debe de mostrarse correctamente con valores por defecto', () => {
        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
    });

    test('debe de mostrarse a batman y el input con el valor del queryString', () => {
        render(
            <AuthContext.Provider value={contextValue}>
                <MemoryRouter initialEntries={['/search?q=batman']}>
                    <PrivateRoute>
                        <HeroesRoutes />
                    </PrivateRoute>
                </MemoryRouter>
            </AuthContext.Provider>
        );

        const input = screen.getByTestId('input-search');
        const img = screen.getByTestId('hero-image');

        expect(input.value).toBe('batman');
        expect(img.src).toContain('/heroes/dc-batman.jpg');
    });

    test('debe de mostrarse el alert de buscar un heroe', () => {
        render(
            <AuthContext.Provider value={contextValue}>
                <MemoryRouter initialEntries={['/search']}>
                    <PrivateRoute>
                        <HeroesRoutes />
                    </PrivateRoute>
                </MemoryRouter>
            </AuthContext.Provider>
        );

        const alertSearch = screen.getByTestId('search-info');

        expect(alertSearch).toBeTruthy();
    });

    test('debe de mostrarse el alert de no encontrado', () => {
        render(
            <AuthContext.Provider value={contextValue}>
                <MemoryRouter initialEntries={['/search?q=myheroe']}>
                    <PrivateRoute>
                        <HeroesRoutes />
                    </PrivateRoute>
                </MemoryRouter>
            </AuthContext.Provider>
        );

        const alertNotFound = screen.getByTestId('notfound-info');

        expect(alertNotFound).toBeTruthy();
    });

    test('debe de llamar el navigate a la pantalla nueva', () => {
        render(
            <AuthContext.Provider value={contextValue}>
                <MemoryRouter initialEntries={['/search']}>
                    <PrivateRoute>
                        <HeroesRoutes />
                    </PrivateRoute>
                </MemoryRouter>
            </AuthContext.Provider>
        );

        const input = screen.getByTestId('input-search');
        const button = screen.getByTestId('search-btn');

        act(() => {

            // changing the value of the input field
            fireEvent.change(input, { target: { value: 'superman' } });

            fireEvent.click(button);
        })

        expect( mockUseNavigate ).toHaveBeenCalledWith(`/search?q=${input.value}`);
    })
});