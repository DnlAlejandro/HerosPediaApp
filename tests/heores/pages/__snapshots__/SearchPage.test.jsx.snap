// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en <SearchPage/> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1
    class="text-3xl font-bold text-indigo-600"
  >
    Results
  </h1>
  <div
    class="mt-5 bg-blue-100 p-5 rounded-lg text-blue-600 border border-blue-300/40 flex items-center gap-3 "
    data-testid="search-info"
  >
    <svg
      class="w-6 h-6"
      fill="none"
      stroke="currentColor"
      stroke-width="1.5"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m21 21-5.197-5.197m0 0A7.5 7.5 0 1 0 5.196 5.196a7.5 7.5 0 0 0 10.607 10.607Z"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
    <p
      class="font-light"
    >
      Search a hero
    </p>
  </div>
  <div
    class="flex gap-4 flex-wrap"
  />
</div>
`;
